perun-core/src/main/java/cz/metacentrum/perun/core/entry/ResourcesManagerEntry.java

```diff
@@ -99,6 +99,41 @@ public Resource createResource(PerunSession sess, Resource resource, Vo vo, Faci
 		return getResourcesManagerBl().createResource(sess, resource, vo, facility);
 	}
 
+	public Resource copyResource(PerunSession sess, Resource templateResource, Resource destinationResource, boolean withGroups) throws InternalErrorException, ResourceNotExistsException, FacilityNotExistsException, PrivilegeException, VoNotExistsException, ResourceExistsException {
+		Utils.checkPerunSession(sess);
+		Utils.notNull(templateResource, "Template Resource");
+		Utils.notNull(destinationResource, "Destination Resource");
+		getResourcesManagerBl().checkResourceExists(sess, templateResource);
+
+		Facility destinationFacility = getPerunBl().getFacilitiesManagerBl().getFacilityById(sess, destinationResource.getFacilityId());
+		Facility templateResourceFacility = getPerunBl().getFacilitiesManagerBl().getFacilityById(sess, templateResource.getFacilityId());
+
+		//check if user is facility admin of template resource's facility
+		if (!AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, templateResourceFacility)) {
+			throw new PrivilegeException(sess, "User is not facility admin of template Resource's facility.");
+		}
+
+		//check if user is facility admin of destination's facility
+		if (!AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, destinationFacility)) {
+			throw new PrivilegeException(sess, "User is not facility admin of destination Resource's facility.");
+		}
+
+		if(withGroups) {
+			Vo destinationVo = getPerunBl().getVosManagerBl().getVoById(sess, destinationResource.getVoId());
+			Vo templateVo = getPerunBl().getVosManagerBl().getVoById(sess, templateResource.getVoId());
+
+			if(!destinationVo.equals(templateVo)) {
```

Comment:
You don't need to get whole VO objects from Perun. You can just check ids of them there and for authorization can be used object resource instead of VO.